### PR TITLE
[glog] update to v0.5

### LINKED ISSUE #17731: [glog] update to 0.5.0
Library name: glog

New version number: https://github.com/google/glog/releases/tag/v0.5.0

Other information that may be useful (release notes, etc...)

### PR DESCRIPTION
**Pull request by jiayuehua**
**Describe the pull request**
glog update to new version v0.5
- #### What does your PR fix?  
  Fixes #...

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  <all >, <No>

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  <Yes>

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/glog/glog_disable_debug_postfix.patch [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
-@@ -44,7 +44,7 @@ include (CTest)
- include (DetermineGflagsNamespace)
- include (GNUInstallDirs)
- 
--set (CMAKE_DEBUG_POSTFIX d)
-+#set (CMAKE_DEBUG_POSTFIX d)
- set (CMAKE_THREAD_PREFER_PTHREAD 1)
- 
- if (WITH_GFLAGS)
+diff --git a/CMakeLists.txt b/CMakeLists.txt
```
Comment (JonLiu1993) [inline ports/glog/glog_disable_debug_postfix.patch:1]: @jiayuehua ,Thanks for your pr, Could you please keep the original format of EOL for this patch, maybe LF?

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Comment by JonLiu1993** [comment]
Looks CI tested failed on many triplet, this is error log, could you please take a look:
https://dev.azure.com/vcpkg/public/_build/results?buildId=64831&view=artifacts&pathAsName=false&type=publishedArtifacts

File: ports/glog/glog_disable_debug_postfix.patch (hunk 1 above)
Comment (jiayuehua) [reply ports/glog/glog_disable_debug_postfix.patch:1]: ok

**Comment by jiayuehua** [comment]
> Looks CI tested failed on many triplet, this is error log, could you please take a look: [dev.azure.com/vcpkg/public/_build/results?buildId=64831&view=artifacts&pathAsName=false&type=publishedArtifacts](https://dev.azure.com/vcpkg/public/_build/results?buildId=64831&view=artifacts&pathAsName=false&type=publishedArtifacts)



> Looks CI tested failed on many triplet, this is error log, could you please take a look: [dev.azure.com/vcpkg/public/_build/results?buildId=64831&view=artifacts&pathAsName=false&type=publishedArtifacts](https://dev.azure.com/vcpkg/public/_build/results?buildId=64831&view=artifacts&pathAsName=false&type=publishedArtifacts)

ok

**Comment by jiayuehua** [comment]
@JonLiu1993 I have commit my fix. Now linux ci can run properly. And windows fail message is not related to glog.

**Comment by JonLiu1993** [comment]
> @JonLiu1993 I have commit my fix. Now linux ci can run properly. And windows fail messages is not related to glog.
@jiayuehua ,Judging from the CI error log, there are three ports that rely on glog and all have error. 
They are evpp:
[install-x86-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/7772824/install-x86-windows-dbg-out.log)
cartographer:
[install-x86-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/7772832/install-x86-windows-dbg-out.log)
folly:
[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/7772840/install-x64-windows-dbg-out.log)

**Comment by jiayuehua** [comment]
D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1088): error C3861: '__sync_add_and_fetch': identifier not found
D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1088): error C3861: '__sync_sub_and_fetch': identifier not found
D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1126): error C3861: '__sync_add_and_fetch': identifier not found
D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1126): error C3861: '__sync_sub_and_fetch': identifier not found
folly error is folly self's bug. and evpp catographer not report compile error and just failed. I think It maybe because ci machine lack memory to run that much build process.

**Comment by JonLiu1993** [comment]
> D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1088): error C3861: '__sync_add_and_fetch': identifier not found D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1088): error C3861: '__sync_sub_and_fetch': identifier not found D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1126): error C3861: '__sync_add_and_fetch': identifier not found D:\buildtrees\folly\src\1.06.14.00-b189899f14.clean\folly\io\async\AsyncServerSocket.cpp(1126): error C3861: '__sync_sub_and_fetch': identifier not found folly error is folly self's bug. and evpp catographer not report compile error and just failed. I think It maybe because ci machine lack memory to run that much build processes.

Thank you for your investigation. There may also be bugs in glog itself. Some users encountered some problems when updating glog. You can refer to this pr #20351 to see if these bugs can be solved.

**Comment by jiayuehua** [comment]
> __sync_sub_and_fetch

It's glog issue, not folly.

**Comment by jiayuehua** [comment]
@JonLiu1993 I have fix it those days.

File: ports/glog/portfile.cmake [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
+    SHA512 445e4338f3d81cd0b065f2da9c6ce343c243263ca144cea424ef97531a4e9e09c06ffd6942ac01c5213a8003c75cfbbede3c4028d12f0134f23ff29314769c1a
     HEAD_REF master
     PATCHES
        glog_disable_debug_postfix.patch
+       fix_glog_CMAKE_MODULE_PATH.patch
+       platforminfo.patch
+       fix_glog_config.h.patch
 )
 
 vcpkg_configure_cmake(
```
Comment (JonLiu1993) [inline ports/glog/portfile.cmake:14]:
```suggestion
vcpkg_cmake_configure(
```

File: ports/glog/portfile.cmake [hunk 3]
```diff
@@ -19,8 +21,9 @@ vcpkg_install_cmake()
 
 vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/glog)
```
Comment (JonLiu1993) [inline ports/glog/portfile.cmake:22]:
```suggestion
vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/glog)
```

File: ports/glog/vcpkg.json [hunk 4]
```diff
@@ -1,7 +1,6 @@
 {
   "name": "glog",
-  "version-string": "0.4.0",
-  "port-version": 4,
+  "version": "0.5.0",
   "description": "C++ implementation of the Google logging module",
   "homepage": "https://github.com/google/glog",
   "dependencies": [
```
Comment (JonLiu1993) [inline ports/glog/vcpkg.json:6]:
```suggestion
  "dependencies": [
      {
          "name": "vcpkg-cmake",
          "host": true
      },
      {
          "name": "vcpkg-cmake-config",
          "host": true
      },
```

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Comment by BillyONeal** [comment]
Can you explain what these patches are doing and/or have they been submitted upstream? In particular it seems odd to duplicate `config.h` and then insert the only users of that...

**Comment by jiayuehua** [comment]
Billy. glog-modules.cmake.in will generate glog-modules.cmake and pollute CMAKE_MODULE_PATH. That make many other port build fail. And Vcpkg only use config type find_package, So I delete all cmake code related to glog-modules.cmake.
config.h Is needed because it contains plaform specific infomation. And origin glog don't install it . If not install, many ports on windows will build fail. And I add it to glog/config.h other then config.h. Because its a header file that belongs to glog. Glog self test use "config.h" , So that I need to config again.

**Comment by BillyONeal** [comment]
>I don't summit it to upstream, because this version is alread released nearly one year.

The reason it's important to submit changes to upstream like this are:

1. We might be modifying the library in a way upstream may not like :)
2. We want the *next* version to have incorporated these patches so that we don't need to reinvestigate/regneerate/reunderstand what the patch is doing every time there's a version update.

>Billy. glog-modules.cmake.in will generate glog-modules.cmake and pollute CMAKE_MODULE_PATH. That make many other port build fail. And Vcpkg only use config type find_package, So I delete all cmake code related to glog-modules.cmake.

Yes, that looked like it might be "specific to vcpkg" stuff, so it would make sense to not submit that upstream. That's why I was in particular asking about the config.h stuff.

>config.h Is needed because it contains plaform specific infomation. And origin glog don't install it . If not install, many ports on windows will build fail. And I add it to glog/config.h other then config.h. Because its a header file that belongs to glog. Glog self test use "config.h" , So that I need to config again.

I'm very confused. Are you saying upstream is always broken? How is config.h expected to be present in normal conditions?

**Comment by jiayuehua** [comment]
I will see if upstream already fix it.

Upstream is always broken.  12 days ago, as jonliu said, CI prove that:
"Judging from the CI error log, there are three ports that rely on glog and all have error.
They are evpp:
install-x86-windows-dbg-out.log
cartographer:
install-x86-windows-dbg-out.log
folly:
install-x64-windows-dbg-out.log ". 
Must include glog/config.h in glog/logging.h otherwise those port can't build successfully.

**Comment by BillyONeal** [comment]
Can you link to upstream where this being discussed?

**Comment by jiayuehua** [comment]
billy,upstream say they fix it 3 month ago on master branch. I think since v0.5.0 is just a git tag, can't be fixed.

**Comment by BillyONeal** [comment]
Great! Do you have a link to the discussion so we can have it on record next time we go to update this port? Thanks!

**Comment by jiayuehua** [comment]
Billy, Three days ago  I link a discussion named "v0.5.0 LOG_EVERY...". inside this one.

**Comment by JonLiu1993** [comment]
> Billy, Three days ago I link the discussion that named "v0.5.0 LOG_EVERY...". inside this one.

@jiayuehua ,Solved by this pr? https://github.com/google/glog/pull/716

**Comment by jiayuehua** [comment]
@JonLiu1993  Upstream say the fix Should be : https://github.com/google/glog/pull/716. I haven't checked.

**Comment by BillyONeal** [comment]
> Billy, Three days ago I link the discussion that named "v0.5.0 LOG_EVERY...". inside this one.

I see, I either didn't see that or was confused since there doesn't seem to be a direct link between the discussion there and specifically installing `config.h`.

Considering upstream seems to have chosen a different solution to the problem I think we need to at least try using their change as indicated by @JonLiu1993

**Comment by jiayuehua** [comment]
I will have a look at upstream solution, Billy.

**Comment by LilyWangLL** [comment]
> I will have a look at upstream solution, Billy.

https://github.com/google/glog/pull/716 was merged after new release version v0.5. You can add this fix by download it from https://patch-diff.githubusercontent.com/raw/google/glog/pull/716.diff at runtime to verify this soluction is useful or not?

**Comment by LilyWangLL** [comment]
Ping @jiayuehua for response. Is work still being done for this PR?

**Comment by jiayuehua** [comment]
ok I will have a look soon.

**Comment by jiayuehua** [comment]
Upstream fix solution is useful. Verified. I will commit patch soon just.

**Comment by jiayuehua** [comment]
@LilyWangLL I has upload my fix, CI microsoft.vcpkg.pr report error, But I can't see the cmake build error log. Can you help me download it? My local machine those ports like evpp can build successfully.
image

**Comment by JonLiu1993** [comment]
@jiayuehua ,This is error log, please take a view:
[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/8183554/install-x64-windows-dbg-out.log)

**Comment by jiayuehua** [comment]
> @jiayuehua ,This is error log, please take a view: [install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/8183554/install-x64-windows-dbg-out.log)

@JonLiu1993 thanks.

**Comment by JonLiu1993** [comment]
@jiayuehua, Is work still being done for this PR?

**Comment by jiayuehua** [comment]
yes,  my daily job is a bit busy. every end of the season busy. I will fing time to see it soon.

**Comment by LilyWangLL** [comment]
I convert this PR to draft, if this issue ready for review, please ping for us.

**Comment by draveness** [comment]
I know this is complicated, but I have to say that I opened the issue almost one year ago 🤣 

ref: https://github.com/microsoft/vcpkg/issues/17731

**Comment by jiayuehua** [comment]
I will see in the coming days.

**Comment by jiayuehua** [comment]
ref to https://rpg.ifi.uzh.ch/docs/glog.html last part , I have fixed rsocket. @LilyWangLL

**Comment by LilyWangLL** [comment]
@jiayuehua The new error on CI pipeline, you can download error log from: https://dev.azure.com/vcpkg/public/_build/results?buildId=70251&view=artifacts&pathAsName=false&type=publishedArtifacts.

**Comment by jiayuehua** [comment]
I will have a look.

**Comment by BillyONeal** [comment]
```
D:\buildtrees\caffe2\src\639a981b79-8e24fb82ee.clean\caffe2/utils/conversions.h(168): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
D:\buildtrees\caffe2\src\639a981b79-8e24fb82ee.clean\caffe2/utils/conversions.h(168): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
D:\buildtrees\caffe2\src\639a981b79-8e24fb82ee.clean\caffe2/utils/conversions.h(174): error C2912: explicit specialization 'caffe2::float16 caffe2::convert::To(const float)' is not a specialization of a function template
D:\buildtrees\caffe2\src\639a981b79-8e24fb82ee.clean\caffe2/utils/conversions.h(191): error C2912: explicit specialization 'float caffe2::convert::To(const caffe2::float16)' is not a specialization of a function template
D:\buildtrees\caffe2\src\639a981b79-8e24fb82ee.clean\caffe2/utils/conversions.h(206): error C2912: explicit specialization 'float caffe2::convert::To(const float)' is not a specialization of a function template
```

https://github.com/facebookarchive/caffe2/blob/32f023fe8c0a0327f8f14b1c041536a7c6b1f4ec/caffe2/utils/conversions.h#L168

It looks like maybe before float16 was coming from glog?

**Comment by jiayuehua** [comment]
@BillyONeal It's because Macro "IN" "OUT" introduced by include <windows.h> in glog. "IN" and "OUT" are used as template parameter in caffe2. Marco is evil.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by BillyONeal** [comment]
> @BillyONeal It's because Macro "IN" "OUT" introduced by include <windows.h> in glog. "IN" and "OUT" are used as template parameter in caffe2. Marco is evil.

Surprising! IN and OUT aren't used on the lines indicated in the message

**Comment by strega-nil-ms** [comment]
Thanks!

**Comment by jiayuehua** [comment]
Used by the primary template https://github.com/facebookarchive/caffe2/blob/32f023fe8c0a0327f8f14b1c041536a7c6b1f4ec/caffe2/utils/conversions.h#L154. The error message is not so intuitive.

**Comment by whuaegeanse** [comment]
@jiayuehua Applying nogdi-nominmax.patch makes a conficts with freeimage because of using macro NOGDI. 
https://github.com/colmap/colmap/issues/1511

**Pull request merged**